README.md
# Sharetribe Flex Template for Web

[![CircleCI](https://circleci.com/gh/sharetribe/ftw-product.svg?style=svg)](https://circleci.com/gh/sharetribe/ftw-product)

This is a template web application for a Sharetribe Flex marketplace ready to be extended and
customized. It is based on an application bootstrapped with
[create-react-app](https://github.com/facebookincubator/create-react-app) with some additions,
namely server side rendering and a custom CSS setup.

> Note: We also have [FTW-daily](https://github.com/sharetribe/ftw-daily) for day-based processes
> and [FTW-hourly](https://github.com/sharetribe/ftw-hourly) for time-based processes. You can learn
> more on
> [Flex Docs](https://www.sharetribe.com/docs/ftw-introduction/how-to-customize-ftw/#choose-a-template)

## Quick start

If you just want to get the app running quickly to test it out, first install
[Node.js](https://nodejs.org/) and [Yarn](https://yarnpkg.com/), and follow along:

```sh
git clone git@github.com:sharetribe/ftw-product.git  # clone this repository
cd ftw-product/                                      # change to the cloned directory
yarn install                                         # install dependencies
yarn run config                                      # add the mandatory env vars to your local config
yarn run dev                                         # start the dev server, this will open a browser in localhost:3000
```

You can also follow along the
[Getting started with FTW](https://www.sharetribe.com/docs/tutorials/getting-started-with-ftw/)
tutorial in the [Flex Docs website](https://www.sharetribe.com/docs/).

For more information of the configuration, see the
[FTW Environment configuration variables](https://www.sharetribe.com/docs/references/ftw-env/)
reference in Flex Docs.

### For Windows users

Change `export` to `set` in the package.json file if you're using Windows/DOS. You need to do the
change to "dev" and "dev-sever" commands.

```
"dev": "yarn run config-check&&set NODE_ENV=development&& set REACT_APP_DEV_API_SERVER_PORT=3500&&concurrently --kill-others \"yarn run dev-frontend\" \"yarn run dev-backend\""
```

```
"dev-server": "set NODE_ENV=development&& set PORT=4000&& set REACT_APP_CANONICAL_ROOT_URL=http://localhost:4000&&yarn run build&&nodemon --watch server server/index.js"
```

We strongly recommend installing
[Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about), if you are
developing on Windows. These templates are made for Unix-like web services which is the most common
environment type on host-services for web apps. Also, Flex Docs uses Unix-like commands in articles
instead of DOS commands.

## Getting started with your own customization

If you want to build your own Flex marketplace by customizing the template application, see the
[How to Customize FTW](https://www.sharetribe.com/docs/guides/how-to-customize-ftw/) guide in Flex
Docs.

## Deploying to Heroku

**Note:** Remember to fork the repository before deploying the application. Connecting your own
Github repository to Heroku will make manual deploys easier.

See the
[How to deploy FTW to production](https://www.sharetribe.com/docs/guides/how-to-deploy-ftw-to-production/)
guide in Flex Docs for more information.

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

## Documentation

See the Flex Docs site: https://www.sharetribe.com/docs/

## Get help – join Sharetribe Flex Developer Slack channel

If you have any questions about development, the best place to ask them is the Flex Developer Slack
channel at https://www.sharetribe.com/flex-slack

## License

This project is licensed under the terms of the Apache-2.0 license.

See [LICENSE](LICENSE)


CHANGELOG.md
<<<<<<< HEAD
# Change Log

We are not following semantic versioning in this template app since any change could potentially be
a breaking change for forked customization projects. We are still experimenting with what is a good
way to update this template, but currently, we follow a pattern:

- Major version change (v**X**.0.0): Changes to several pages and other components. Consider
  implementing this without merging upstream (we'll provide instructions).
- Minor version change (v0.**X**.0): New features and changes to a single page. These are likely to
  cause conflicts.
- Patch (v0.0.**X**): Bug fixes and small changes to components.

---

## Upcoming version 2022-XX-XX

## [v10.0.1] 2022-07-11

- ListingPage.test.js needed an update due to missing 'stock' in includes.

  [v10.0.1]: https://github.com/sharetribe/ftw-product/compare/v10.0.0...v10.0.1

## [v10.0.0] 2022-07-11

This is a major release since it includes a major update to sharetribe-scripts aka our fork of
Create React App.It includes major update to Webpack (v4 > v5) and PostCSS (v7 > v8). These caused
some advanced CSS syntax to be invalid - and therefore some changes must be done to CSS files.

- CSS Property Sets are deprecated and the related file is removed from the codebase
- Custom media queries file need to be imported in all the files, that use them.

Read more from PR: https://github.com/sharetribe/ftw-daily/pull/1531

Changes:

- [change] sharetribe-scripts is updated to v6.0.0. This causes a new major release for FTW
  templates. Because most of the CSS files need to be updated!
  [#1531](https://github.com/sharetribe/ftw-daily/pull/1531)

- [change] ImageCarousel uses react-image-gallery. It renders all the images early on and supports
  touch swipes. [#1529](https://github.com/sharetribe/ftw-daily/pull/1529)
- [fix] Fixing a small typo. [#1518](https://github.com/sharetribe/ftw-daily/pull/1518)

  [v10.0.0]: https://github.com/sharetribe/ftw-product/compare/v9.3.1...v10.0.0

## [v9.3.1] 2022-05-19

- [fix] undefined REACT_APP_GOOGLE_ANALYTICS_ID caused an error.
  [#150](https://github.com/sharetribe/ftw-product/pull/150)

  [v9.3.0]: https://github.com/sharetribe/ftw-product/compare/v9.3.0...v9.3.1

## [v9.3.0] 2022-05-19

Note: this caused an error. Use v9.3.1 instead.

- [fix] undefined REACT_APP_GOOGLE_ANALYTICS_ID caused an error.
  [#149](https://github.com/sharetribe/ftw-product/pull/149)

### Updates from upstream (FTW-daily v8.6.0)

- [change] Google Analytics: remove Universal Analytics and start supporting GA4.

  NOTE: you need to update the Google Analytics id to GA4's id (starting with 'G-' prefix).

  [#1508](https://github.com/sharetribe/ftw-daily/pull/1508)

- [change] Update some outdated dependencies.
  [#1514](https://github.com/sharetribe/ftw-daily/pull/1514)

  [v9.3.0]: https://github.com/sharetribe/ftw-product/compare/v9.2.0...v9.3.0

## [v9.2.0] 2022-05-16

- Add support for hosted translations. (PR made in upstream repo: FTW-daily)

- [delete] Remove old unused translation keys
  [#146](https://github.com/sharetribe/ftw-product/pull/146)

### Updates from upstream (FTW-daily v8.5.0)

- [add] Add support for hosted translations.

  - This PR fetches "content/translation.json" from a new Asset Delivery API. The file is editable
    through the Flex Console.
  - It also adds all the missing translation keys to existing non-English translation files. This
    means that those files might now include messages in English.

  [#1510](https://github.com/sharetribe/ftw-daily/pull/1510)

- [delete] Remove old unused translation keys.
  [#1511](https://github.com/sharetribe/ftw-daily/pull/1511)

  [v9.2.0]: https://github.com/sharetribe/ftw-daily/compare/v9.1.2...v9.2.0

## [v9.1.2] 2022-03-16

- [fix] Process graph had an error: provider notification about dispute being canceled was not sent
  to provider but customer. [#135](https://github.com/sharetribe/ftw-product/pull/135)
- [fix] CheckoutPage: add missing link for fallback error, when submitting.
  [#137](https://github.com/sharetribe/ftw-product/pull/137)
- [fix] CheckoutPage: if the saveAfterOnetimePayment checkbox is checked and then unchecked, the
  value is already an array. [#136](https://github.com/sharetribe/ftw-product/pull/136)

### Updates from upstream (FTW-daily v8.4.1)

- [fix] Portal version of Modals couldn't use 100vh on mobile Safari. Safari changes the window's
  height based on if the location bar is squeezed or not (and that depends on scroll-effect).
  [#1501](https://github.com/sharetribe/ftw-daily/pull/1501)
- [fix] Setting multiple environment variables in Windows requires using the set command before each
  individual variable. Updated the 'for windows users' section in documentation.
  [#1491](https://github.com/sharetribe/ftw-daily/pull/1491)
- [add] Code comment about "REACT_APP" prefix in environment variables.
  [#1492](https://github.com/sharetribe/ftw-daily/pull/1492)

  [v9.1.2]: https://github.com/sharetribe/ftw-product/compare/v9.1.1...v9.1.2

## [v9.1.1] 2022-01-24

- [fix] A full-page refresh on product when stock = 1 causes 500 error. SSR gets change request
  automatically from Final Form/FromSpy, and therefore window.fetch gets called on the server.
  [#133](https://github.com/sharetribe/ftw-product/pull/133)
- [fix] Add a missing return statement on ProductOrderForm. The code path is relevant only for
  listings that are created outside of EditListingWizard.
  [#127](https://github.com/sharetribe/ftw-product/pull/127)

  [v9.1.1]: https://github.com/sharetribe/ftw-product/compare/v9.1.0...v9.1.1

## [v9.1.0] 2021-12-02

### Updates from upstream (FTW-daily v8.4.0)

- [fix] Mobile safari zooms if input-related elements have smaller font-size than 16px. This updates
  textarea and select element styles too. [#1489](https://github.com/sharetribe/ftw-daily/pull/1489)

server/index.js
/**
 * This is the main server to run the production application.
 *
 * Running the server requires that `npm run build` is run so that the
 * production JS bundle can be imported.
 *
 * This server renders the requested URL in the server side for
 * performance, SEO, etc., and properly handles redirects, HTTP status
 * codes, and serving the static assets.
 *
 * When the application is loaded in a browser, the client side app
 * takes control and all the functionality such as routing is handled
 * in the client.
 */

// This enables nice stacktraces from the minified production bundle
require('source-map-support').install();

// Configure process.env with .env.* files
require('./env').configureEnv();

const http = require('http');
const https = require('https');
const express = require('express');
const helmet = require('helmet');
const compression = require('compression');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');
const enforceSsl = require('express-enforces-ssl');
const path = require('path');
const sharetribeSdk = require('sharetribe-flex-sdk');
const sitemap = require('express-sitemap');
const passport = require('passport');
const auth = require('./auth');
const apiRouter = require('./apiRouter');
const wellKnownRouter = require('./wellKnownRouter');
const { getExtractors } = require('./importer');
const renderer = require('./renderer');
const dataLoader = require('./dataLoader');
const fs = require('fs');
const log = require('./log');
const { sitemapStructure } = require('./sitemap');
const csp = require('./csp');
const sdkUtils = require('./api-util/sdk');
require("dotenv").config()
const buildPath = path.resolve(__dirname, '..', 'build');
const env = process.env.REACT_APP_ENV;
const dev = process.env.REACT_APP_ENV === 'development';
const PORT = parseInt(process.env.PORT, 10);
const CLIENT_ID = process.env.REACT_APP_SHARETRIBE_SDK_CLIENT_ID || "0af5ed88-4ae8-4b52-806e-b3c501011ab5";
const BASE_URL = process.env.REACT_APP_SHARETRIBE_SDK_BASE_URL;
const ASSET_CDN_BASE_URL = process.env.REACT_APP_SHARETRIBE_SDK_ASSET_CDN_BASE_URL;
const TRANSIT_VERBOSE = process.env.REACT_APP_SHARETRIBE_SDK_TRANSIT_VERBOSE === 'true';
const USING_SSL = process.env.REACT_APP_SHARETRIBE_USING_SSL === 'true';
const TRUST_PROXY = process.env.SERVER_SHARETRIBE_TRUST_PROXY || null;
const CSP = process.env.REACT_APP_CSP;
const cspReportUrl = '/csp-report';
const cspEnabled = CSP === 'block' || CSP === 'report';
const app = express();

const errorPage = fs.readFileSync(path.join(buildPath, '500.html'), 'utf-8');

// load sitemap and robots file structure
// and write those into files
sitemap(sitemapStructure()).toFile();

// Setup error logger
log.setup();
// Add logger request handler. In case Sentry is set up
// request information is added to error context when sent
// to Sentry.
app.use(log.requestHandler());

// The helmet middleware sets various HTTP headers to improve security.
// See: https://www.npmjs.com/package/helmet
// Helmet 4 doesn't disable CSP by default so we need to do that explicitly.
// If csp is enabled we will add that separately.

app.use(
  helmet({
    contentSecurityPolicy: false,
  })
);

if (cspEnabled) {
  // When a CSP directive is violated, the browser posts a JSON body
  // to the defined report URL and we need to parse this body.
  app.use(
    bodyParser.json({
      type: ['json', 'application/csp-report'],
    })
  );

  // CSP can be turned on in report or block mode. In report mode, the
  // browser checks the policy and calls the report URL when the
  // policy is violated, but doesn't block any requests. In block
  // mode, the browser also blocks the requests.

  // In Helmet 4,supplying functions as directive values is not supported.
  // That's why we need to create own middleware function that calls the Helmet's middleware function
  const reportOnly = CSP === 'report';
  app.use((req, res, next) => {
    csp(cspReportUrl, USING_SSL, reportOnly)(req, res, next);
  });
}

// Redirect HTTP to HTTPS if USING_SSL is `true`.
// This also works behind reverse proxies (load balancers) as they are for example used by Heroku.
// In such cases, however, the TRUST_PROXY parameter has to be set (see below)
//
// Read more: https://github.com/aredo/express-enforces-ssl
//
if (USING_SSL) {
  app.use(enforceSsl());
}

// Set the TRUST_PROXY when running the app behind a reverse proxy.
//
// For example, when running the app in Heroku, set TRUST_PROXY to `true`.
//
// Read more: https://expressjs.com/en/guide/behind-proxies.html
//
if (TRUST_PROXY === 'true') {
  app.enable('trust proxy');
} else if (TRUST_PROXY === 'false') {
  app.disable('trust proxy');
} else if (TRUST_PROXY !== null) {
  app.set('trust proxy', TRUST_PROXY);
}

app.use(compression());
app.use('/static', express.static(path.join(buildPath, 'static')));
// server robots.txt from the root
app.use('/robots.txt', express.static(path.join(buildPath, 'robots.txt')));
app.use(cookieParser());

// These .well-known/* endpoints will be enabled if you are using FTW as OIDC proxy
// https://www.sharetribe.com/docs/cookbook-social-logins-and-sso/setup-open-id-connect-proxy/
// We need to handle these endpoints separately so that they are accessible by Flex
// even if you have enabled basic authentication e.g. in staging environment.
app.use('/.well-known', wellKnownRouter);

// Use basic authentication when not in dev mode. This is
// intentionally after the static middleware and /.well-known
// endpoints as those will bypass basic auth.
if (!dev) {
  const USERNAME = process.env.BASIC_AUTH_USERNAME;
  const PASSWORD = process.env.BASIC_AUTH_PASSWORD;
  const hasUsername = typeof USERNAME === 'string' && USERNAME.length > 0;
  const hasPassword = typeof PASSWORD === 'string' && PASSWORD.length > 0;

  // If BASIC_AUTH_USERNAME and BASIC_AUTH_PASSWORD have been set - let's use them
  if (hasUsername && hasPassword) {
    app.use(auth.basicAuth(USERNAME, PASSWORD));
  }
}

// Initialize Passport.js  (http://www.passportjs.org/)
// Passport is authentication middleware for Node.js
// We use passport to enable authenticating with
// a 3rd party identity provider (e.g. Facebook or Google)
app.use(passport.initialize());

// Server-side routes that do not render the application
app.use('/api', apiRouter);

const noCacheHeaders = {
  'Cache-control': 'no-cache, no-store, must-revalidate',
};

// Instantiate HTTP(S) Agents with keepAlive set to true.
// This will reduce the request time for consecutive requests by
// reusing the existing TCP connection, thus eliminating the time used
// for setting up new TCP connections.
const httpAgent = new http.Agent({ keepAlive: true });
const httpsAgent = new https.Agent({ keepAlive: true });

app.get('*', (req, res) => {
  if (req.url.startsWith('/static/')) {
    // The express.static middleware only handles static resources
    // that it finds, otherwise passes them through. However, we don't
    // want to render the app for missing static resources and can
    // just return 404 right away.
    return res.status(404).send('Static asset not found.');
  }

  if (req.url === '/_status.json') {
    return res.status(200).send({ status: 'ok' });
  }

  const context = {};

  // Get handle to tokenStore
  // We check in unauthorized cases if requests have set tokens to cookies
  const tokenStore = sharetribeSdk.tokenStore.expressCookieStore({
    clientId: CLIENT_ID,
    req,
    res,
    secure: USING_SSL,
  });

  const baseUrl = BASE_URL ? { baseUrl: BASE_URL } : {};
  const assetCdnBaseUrl = ASSET_CDN_BASE_URL ? { assetCdnBaseUrl: ASSET_CDN_BASE_URL } : {};

  const sdk = sharetribeSdk.createInstance({
    transitVerbose: TRANSIT_VERBOSE,
    clientId: CLIENT_ID,
    httpAgent: httpAgent,
    httpsAgent: httpsAgent,
    tokenStore,
    typeHandlers: sdkUtils.typeHandlers,
    ...baseUrl,
    ...assetCdnBaseUrl,
  });

  // Until we have a better plan for caching dynamic content and we
  // make sure that no sensitive data can appear in the prefetched
  // data, let's disable response caching altogether.
  res.set(noCacheHeaders);

  // Get chunk extractors from node and web builds
  // https://loadable-components.com/docs/api-loadable-server/#chunkextractor
  const { nodeExtractor, webExtractor } = getExtractors();

  // Server-side entrypoint provides us the functions for server-side data loading and rendering
  const nodeEntrypoint = nodeExtractor.requireEntrypoint();
  const { default: renderApp, ...appInfo } = nodeEntrypoint;

  dataLoader
    .loadData(req.url, sdk, appInfo)
    .then(data => {
      const html = renderer.render(req.url, context, data, renderApp, webExtractor);

      if (dev) {
        const debugData = {
          url: req.url,
          context,
        };
        console.log(`\nRender info:\n${JSON.stringify(debugData, null, '  ')}`);
      }

      if (context.unauthorized) {
        // Routes component injects the context.unauthorized when the
        // user isn't logged in to view the page that requires
        // authentication.
        sdk.authInfo().then(authInfo => {
          if (authInfo && authInfo.isAnonymous === false) {
            // It looks like the user is logged in.
            // Full verification would require actual call to API
            // to refresh the access token
            res.status(200).send(html);
          } else {
            // Current token is anonymous.
            res.status(401).send(html);
          }
        });
      } else if (context.forbidden) {
        res.status(403).send(html);
      } else if (context.url) {
        // React Router injects the context.url if a redirect was rendered
        res.redirect(context.url);
      } else if (context.notfound) {
        // NotFoundPage component injects the context.notfound when a
        // 404 should be returned
        res.status(404).send(html);
      } else {
        res.send(html);
      }
    })
    .catch(e => {
      log.error(e, 'server-side-render-failed');
      res.status(500).send(errorPage);
    });
});

// Set error handler. If Sentry is set up, all error responses
// will be logged there.
app.use(log.errorHandler());

if (cspEnabled) {
  // Dig out the value of the given CSP report key from the request body.
  const reportValue = (req, key) => {
    const report = req.body ? req.body['csp-report'] : null;
    return report && report[key] ? report[key] : key;
  };

  // Handler for CSP violation reports.
  app.post(cspReportUrl, (req, res) => {
    const effectiveDirective = reportValue(req, 'effective-directive');
    const blockedUri = reportValue(req, 'blocked-uri');
    const msg = `CSP: ${effectiveDirective} doesn't allow ${blockedUri}`;
    log.error(new Error(msg), 'csp-violation');
    res.status(204).end();
  });
}

app.listen(PORT || 3000, () => {
  const mode = dev ? 'development' : 'production';
  console.log(`Listening to port ${PORT} in ${mode} mode`);
  if (dev) {
    console.log(`Open http://localhost:${PORT}/ and start hacking!`);
  }
});
